core/schema/modulesource.go

```diff
@@ -946,17 +947,44 @@ func (s *moduleSourceSchema) loadModuleSourceContext(
 		fullIncludePaths = append(fullIncludePaths, src.SourceRootSubpath+"/**/*")
 	}
 
-	fullIncludePaths = append(fullIncludePaths, src.RebasedIncludePaths...)
-
 	switch src.Kind {
 	case core.ModuleSourceKindLocal:
-		err := dag.Select(ctx, dag.Root(), &src.ContextDirectory,
+		moduleSourcePath := src.SourceSubpath
+		if moduleSourcePath == "" {
+			moduleSourcePath = src.SourceRootSubpath
+		}
+
+		// Load .gitignore patterns before loading the module so they can be used in the `Include` argument
+		// instead of Exclude so we get the right behaviour.
+		// We only include the .gitignore from the module source path and above up to the context directory.
+		ignorePatterns, err := loadDirectoryGitIgnorePatterns(ctx, src.Local.ContextDirectoryPath, filepath.Join(src.Local.ContextDirectoryPath, moduleSourcePath))
```

Comment:
Yeah, I can see how this change improves things. @jedevc actually had a good idea to further optimize, which can be done in a follow-up:
- Select all the contents with `host.directory(include=["**/.gitignore"])` and get a `dagql.ObjectResult[*Directory]`
- Use a `Directory.walk(callback)` (doesn't exist yet) to avoid mount/unmounting the directory many times
- Create an internal `Directory._gitignoresFor("path/to/dir")` wrapped in a `DagOpWrapper`

That way if the passed directory is the same the output is cached and nothing will be mounted.

Reply:
Yeah, we absolutely *need* something like this.

Currently, the performance is gonna be crappy because for every gitignore we're going to mount+read+unmount an entire directory using the containerd snapshotter. I imagine this is where the slowdown comes from.

I don't *think* we should block - but this is a pretty urgent follow-up, we shouldn't delay it.